### ISSUE TITLE
Spontaneous 16:Deadline Exceeded after long sleeps

### ISSUE THREAD

**Issue by Nakilon** [GoogleCloudPlatform/google-cloud-ruby#1867]
I have a wrapper. When it faces an exception on calling `Google::Cloud::Logging::Logger`'s `debug`, `info`, `warn`, `error`, methods it just retries:
```ruby
        rescue Google::Cloud::DeadlineExceededError, Google::Cloud::UnauthenticatedError => e
          ruby_logger.error "'#{e}' of #{message.inspect}"
          ruby_logger.info "sleep #{timeout += 1}"
          sleep timeout
          retry
```
It worked fine since spring but nearly a month ago (or a bit earlier) it started to hang -- I mean, sometimes retrying does not help anymore and so it gets stuck in a loop, because I don't want to skip any lines to be logged. So before `sleep timeout` I put a `raise e if timeout > 100` and here is backtrace:
```
E 171203 222609 : '16:Deadline Exceeded'
I 171203 222609 : sleep 98
E 171203 222807 : '16:Deadline Exceeded'
I 171203 222807 : sleep 99
E 171203 223007 : '16:Deadline Exceeded'
I 171203 223007 : sleep 100
E 171203 223222 : '16:Deadline Exceeded'
I 171203 223222 : sleep 101
/home/nakilon/.gem/gems/google-cloud-logging-1.3.1/lib/google/cloud/logging/service.rb:308:in `rescue in execute': 16:Deadline Exceeded (Google::Cloud::UnauthenticatedError)
        from /home/nakilon/.gem/gems/google-cloud-logging-1.3.1/lib/google/cloud/logging/service.rb:305:in `execute'
        from /home/nakilon/.gem/gems/google-cloud-logging-1.3.1/lib/google/cloud/logging/service.rb:104:in `write_entries'
        from /home/nakilon/.gem/gems/google-cloud-logging-1.3.1/lib/google/cloud/logging/project.rb:310:in `write_entries'
        from /home/nakilon/.gem/gems/google-cloud-logging-1.3.1/lib/google/cloud/logging/logger.rb:551:in `write_entry'
        from /home/nakilon/.gem/gems/google-cloud-logging-1.3.1/lib/google/cloud/logging/logger.rb:317:in `add'
        from /home/nakilon/.gem/gems/google-cloud-logging-1.3.1/lib/google/cloud/logging/logger.rb:231:in `warn'
        from /home/nakilon/.gem/bundler/gems/gcplogger-1bb83b0335e7/lib/gcplogger.rb:31:in `call'
        from /home/nakilon/.gem/bundler/gems/gcplogger-1bb83b0335e7/lib/gcplogger.rb:31:in `block (3 levels) in <module:GCPLogger>'
        from main.rb:33:in `block in <main>'
        from main.rb:32:in `loop'
        from main.rb:32:in `<main>'
```
Weird detail is that my program logs a lot of stuff with different log levels but it gets stuck always on the same line:
```
sleep 300
logger.warn "LOOP #{Time.now}"
```
This payload is pretty simple: `"LOOP 2017-12-03 20:30:48 +0000"`
```
$ gem env
RubyGems Environment:
  - RUBYGEMS VERSION: 2.0.14
  - RUBY VERSION: 2.0.0 (2014-11-13 patchlevel 598) [x86_64-linux]
  - INSTALLATION DIRECTORY: /home/nakilon/.gem
  - RUBY EXECUTABLE: /usr/bin/ruby
  - EXECUTABLE DIRECTORY: /home/nakilon/.gem/bin
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /home/nakilon/.gem
     - /home/nakilon/.gem/ruby
     - /usr/share/gems
     - /usr/local/share/gems
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
```
```
$ cat Gemfile.lock
GEM
  remote: https://rubygems.org/
  specs:
    addressable (2.5.2)
      public_suffix (>= 2.0.2, < 4.0)
    faraday (0.13.1)
      multipart-post (>= 1.2, < 3)
    fastimage (1.7.0)
      addressable (~> 2.3, >= 2.3.5)
    google-cloud-core (1.1.0)
      google-cloud-env (~> 1.0)
    google-cloud-env (1.0.1)
      faraday (~> 0.11)
    google-cloud-logging (1.3.1)
      google-cloud-core (~> 1.1)
      google-gax (~> 0.10.1)
      stackdriver-core (~> 1.2)
    google-gax (0.10.1)
      google-protobuf (~> 3.2)
      googleapis-common-protos (>= 1.3.5, < 2.0)
      googleauth (~> 0.6.2)
      grpc (>= 1.7.2, < 2.0)
      rly (~> 0.2.3)
    google-protobuf (3.5.0)
    googleapis-common-protos (1.3.7)
      google-protobuf (~> 3.0)
      googleapis-common-protos-types (~> 1.0)
      grpc (~> 1.0)
    googleapis-common-protos-types (1.0.1)
      google-protobuf (~> 3.0)
    googleauth (0.6.2)
      faraday (~> 0.12)
      jwt (>= 1.4, < 3.0)
      logging (~> 2.0)
      memoist (~> 0.12)
      multi_json (~> 1.11)
      os (~> 0.9)
      signet (~> 0.7)
    grpc (1.7.2)
      google-protobuf (~> 3.1)
      googleapis-common-protos-types (~> 1.0.0)
      googleauth (>= 0.5.1, < 0.7)
    json (2.1.0)
    jwt (1.5.6)
    little-plugger (1.1.4)
    logging (2.2.2)
      little-plugger (~> 1.1)
      multi_json (~> 1.10)
    memoist (0.16.0)
    mini_portile2 (2.1.0)
    multi_json (1.12.2)
    multipart-post (2.0.0)
    nokogiri (1.6.8.1)
      mini_portile2 (~> 2.1.0)
    os (0.9.6)
    public_suffix (2.0.5)
    reddit_bot (1.4.0)
      json
    rly (0.2.3)
    signet (0.8.1)
      addressable (~> 2.3)
      faraday (~> 0.9)
      jwt (>= 1.5, < 3.0)
      multi_json (~> 1.10)
    stackdriver-core (1.2.0)
```
```
$ hostnamectl
   Static hostname: localhost
Transient hostname: instance-1
         Icon name: computer-vm
           Chassis: vm
        Machine ID: a29d1507b8f8596c6e7284b11dc84f8c
           Boot ID: 17531d1bfee14443b222687801a6eea9
    Virtualization: kvm
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-327.36.1.el7.x86_64
      Architecture: x86-64
```

**Comment by quartzmo** [comment]
This is just a shot-in-the-dark suggestion (since I really don't know why the Logging service is requiring that you retry in the first place), but i wonder if it might help to retry with exponential backoff rather than every second.

**Comment by quartzmo** [comment]
Also, do you want to open a Stack Overflow question with the [`google-cloud-logging` tag](https://stackoverflow.com/questions/tagged/google-cloud-logging)? If you do so, please link to it from this issue for future visitors.

**Comment by Nakilon** [comment]
Exponential didn't help:
```
E 171207 040234 : '4:Deadline Exceeded' of "source_ultireddit.size: 100"
I 171207 040241 : sleep 2
E 171207 040451 : '16:Deadline Exceeded' of "source_ultireddit.size: 100"
I 171207 040451 : sleep 4
E 171207 040626 : '16:Deadline Exceeded' of "source_ultireddit.size: 100"
I 171207 040629 : sleep 8
E 171207 040716 : '4:Deadline Exceeded' of "source_ultireddit.size: 100"
I 171207 040716 : sleep 16
E 171207 040804 : '16:Deadline Exceeded' of "source_ultireddit.size: 100"
I 171207 040807 : sleep 32
E 171207 040929 : '16:Deadline Exceeded' of "source_ultireddit.size: 100"
I 171207 040930 : sleep 64
E 171207 041124 : '16:Deadline Exceeded' of "source_ultireddit.size: 100"
I 171207 041124 : sleep 128
E 171207 041404 : '16:Deadline Exceeded' of "source_ultireddit.size: 100"
I 171207 041405 : sleep 256
E 171207 041847 : '16:Deadline Exceeded' of "source_ultireddit.size: 100"
I 171207 041849 : sleep 512
E 171207 042759 : '4:Deadline Exceeded' of "source_ultireddit.size: 100"
I 171207 042804 : sleep 1024
/home/nakilon/.gem/gems/google-cloud-logging-1.3.1/lib/google/cloud/logging/service.rb:308:in `rescue in execute': 4:Deadline Exceeded (Google::Cloud::DeadlineExceededError)
```
However, this time it happened on another line! Seems like it happens in just a random moment of time so larger sleep have larger chance to caught it and so it's unrelated to the specific line of code.
Is there in this gem anything internal such as "connection" to a GCP Logging service? Something that should be kept and automatically renewed but sometimes that renewing fails?

**Comment by Nakilon** [comment]
Here I opened SO question: https://stackoverflow.com/q/47709700/322020

**Comment by quartzmo** [comment]
> Is there in this gem anything internal such as "connection" to a GCP Logging service? Something that should be kept and automatically renewed but sometimes that renewing fails?

I don't think so. The lowest-level stub in library is here:
[google/cloud/logging/v2/logging_service_v2_client.rb](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/google-cloud-logging/v1.3.1/google-cloud-logging/lib/google/cloud/logging/v2/logging_service_v2_client.rb#L217-L225)

**Comment by quartzmo** [comment]
@dazuma Can you suggest anything?

**Comment by dazuma** [comment]
There is a "connection" object internally, but it doesn't need to be renewed. The one thing that does need renewal (sometimes, depending on the setup) is credentials, but that should happen automatically unless you're doing something strange to set up credentials.

Can I ask:

* Are you doing something specific to get credentials (e.g. do you point to a JSON credentials file) or do you use the credentials that GCE gives you automatically?
* In general how much logging do you do? (e.g. a few entries per hour, a few thousand per second, etc.)
* Are you on the stackdriver free tier, or do you have a paid account? (These may have different quotas.)

**Comment by Nakilon** [comment]
* I'm using a keyfile (generated for a service account) like this: 
  ```
  logging = Google::Cloud::Logging.new project: JSON.load(File.read ENV["LOGGING_KEYFILE"])["project_id"]
  machine = [ "gce_instance", {
    "instance_id" => `curl http://metadata.google.internal/computeMetadata/v1/instance/id -H "Metadata-Flavor: Google"`,
    "zone" => `curl http://metadata.google.internal/computeMetadata/v1/instance/zone -H "Metadata-Flavor: Google"`.split("/").last,
  } ]
  logger = Google::Cloud::Logging::Logger.new logging, name, logging.resource(*machine), {}
  ```
  this all is in a wrapper that is used by few other daemons on the same machine and they don't have the issue
* I see around 5000 log items for last 3 days and they are coming in bursts sometimes -- like when I restart the process that is facing the issue
* free tier

Also there is a detail that RAM on my machine is almost fully utilized:
![image](https://user-images.githubusercontent.com/2870363/33963386-ce4510b6-e065-11e7-8e38-52f5e2b646ec.png)
and sometimes responds slowly but:
1) what exception should I get if that's the cause?
2) as I said, other logging processes on the same machine don't have the issue

**Comment by Nakilon** [comment]
Oh and I guess this is another answer on the question about amount of logging:
![image](https://user-images.githubusercontent.com/2870363/33963634-a42842fc-e066-11e7-9ddd-35cec8d57f9f.png)

**Comment by Nakilon** [comment]
@quartzmo the issue happened only once (2 weeks ago) during last months. I think it became rare because I've lowered a bit the memory consuption on my machine. But the gem version I use is now outdated anyway so it may be fixed already, so I don't mind if you close it and maybe some day someone will find this thread and bump it.

**Comment by quartzmo** [comment]
Thanks @Nakilon!

**Comment by qingling128** [comment]
We are seeing this issue as well with fluent-plugin-google-cloud plugin in a cluster with very few logs:
```
2018-10-17 08:56:34 +0000 [warn]: Dropping 206 log message(s) error="16:Deadline Exceeded" error_code="16"
```

Note that error code doesn't match actual error type. See https://github.com/grpc/grpc-go/blob/master/codes/codes.go#L146

16 in fact corresponds to `Unauthenticated`. It could be due to credential expirations.

**Comment by blowmage** [comment]
@qingling128 Can you open a new issue with all the relevant details?
